src/main/kotlin/cacophony/utils/FileInput.kt

```diff
@@ -0,0 +1,91 @@
+package cacophony.utils
+
+import java.io.File
+
+class FileInput(filePath: String) : Input {
+    private val file = File(filePath)
+    private val lines = file.readLines().toMutableList()
+    private val lineBeginPos = mutableListOf<Int>()
+    private var curLineInd = 0
+    private var curPos = 0
+
+    init {
+        // Add new line character to every line (even if it might not have been present in the last one).
+        lines.forEachIndexed { ind, line ->
+            lines[ind] = line + "\n"
+        }
+
+        // Calculate position of each begin of line in the string formed by the concatenation of all lines.
+        // Used to encode Location.
+        var linesLengthSum = 0
+        lineBeginPos.add(0)
+        for (line in lines) {
+            linesLengthSum += line.length
+            lineBeginPos.add(linesLengthSum)
+        }
+    }
+
+    private fun setNextPosition() {
+        if (curLineInd >= lines.size) return
+
+        curPos++
+        if (curPos >= lines[curLineInd].length) {
+            curPos = 0
+            curLineInd++
+        }
+    }
+
+    private fun getCharAtPosition(
+        lineInd: Int,
+        pos: Int,
+    ): Char? {
+        if (lines.elementAtOrNull(lineInd) == null) return null
+        return lines[lineInd].elementAtOrNull(pos)
+    }
+
+    // Location is encoded as the position in the string formed by the concatenation of all lines.
+    private fun encodeLocation(
+        lineInd: Int,
+        pos: Int,
+    ): Location {
+        return Location(lineBeginPos[lineInd] + pos)
+    }
+
+    private fun decodeLocation(loc: Location): Pair<Int, Int> {
```

Comment:
What is someone passes negative or astronomically big index?

Reply:
Fixed, but I think you shouldn't create Location on your own, outside of Input

Reply:
You shouldn't, but what if we add modules and we'll parse two files and somehow mix current locations? Or something other. It doesn't matter that we shouldn't do something, we still should guard ourselves from it.